gtsam_integrator/plugin.cpp

```diff
@@ -20,235 +20,299 @@
+            _pim->resetIntegrationAndSetBias(_imu_bias);
+
+            _navstate_lkf = nav_t {
+                gtsam::Pose3{gtsam::Rot3{imu_int_input.quat}, imu_int_input.position},
+                imu_int_input.velocity
+            };
+        }
+
+        void integrateMeasurement(const imu_t& imu_input, const imu_t& imu_input_next) noexcept
+        {
+            assert(_pim != nullptr && "_pim shuold not be null");
+
+            const gtsam::Vector3 measured_acc{imu_input.am};
+            const gtsam::Vector3 measured_omega{imu_input.wm};
+
+            /// Delta T should be in seconds
+            const double delta_t = imu_input_next.timestamp - imu_input.timestamp;
+
+            _pim->integrateMeasurement(measured_acc, measured_omega, delta_t);
+        }
+
+        bias_t biasHat() const noexcept
+        {
+            assert(_pim != nullptr && "_pim shuold not be null");
+            return _pim->biasHat();
+        }
+
+        nav_t predict() const noexcept
+        {
+            assert(_pim != nullptr && "_pim should not be null");
+            return _pim->predict(_navstate_lkf, _imu_bias);
+        }
+
+    private:
+        bias_t _imu_bias;
+        nav_t _navstate_lkf;
+        pim_ptr_t _pim;
+    };
+
+    std::unique_ptr<PimObject> _pim_obj;
+
+
+    // Remove IMU values older than 'IMU_TTL' from the imu buffer
+    void clean_imu_vec(const double& timestamp) {
+        auto imu_iterator = _imu_vec.begin();
+
+        // Since the vector is ordered oldest to latest, keep deleting until you
+        // hit a value less than 'IMU_TTL' seconds old
+        while (imu_iterator != _imu_vec.end()) {
+            if (timestamp - imu_iterator->timestamp < IMU_TTL) {
+                break;
+            }
+
+            imu_iterator = _imu_vec.erase(imu_iterator);
+        }
+    }
+
+    // Timestamp we are propagating the biases to (new IMU reading time)
+    void propagate_imu_values(const double& timestamp, const time_type& real_time) {
+        auto input_values = _m_imu_integrator_input.get_ro_nullable();
+        if (input_values == nullptr) {
+            return;
+        }
 
 #ifndef NDEBUG
-		if (input_values->last_cam_integration_time > last_cam_time) {
-			std::cout << "New slow pose has arrived!\n";
-			last_cam_time = input_values->last_cam_integration_time;
-		}
+        if (input_values->last_cam_integration_time > last_cam_time) {
+            std::cout << "New slow pose has arrived!\n";
+            last_cam_time = input_values->last_cam_integration_time;
+        }
 #endif
-		pim_->resetIntegrationAndSetBias(imu_bias);
 
-		double time_begin = input_values->last_cam_integration_time + last_imu_offset;
-		double time_end = timestamp + input_values->t_offset;
+        if (_pim_obj == nullptr) {
+            /// We don't have a PimObject -> make and set given the current input
+            _pim_obj = std::make_unique<PimObject>(*input_values);
+            last_imu_offset = input_values->t_offset;
```

Comment:
This update also needs to happen at the end of the function, otherwise `last_imu_offset` is really `first_imu_offset`, and since the VIO updates the offset as time proceeds, it doesn't make sense that we would keep using a stale offset estimate. This is a bug in the previous version too, and also in the rk4 integrator.
But this is also tangential and requires testing of its own, so for now just make a note of this somewhere so we don't forget.